drivers/sensor/bosch/bmm350/bmm350.c

```diff
@@ -0,0 +1,984 @@
+		reg = BMM350_DATA_RATE_12_5HZ;
+	} else if ((odr > 12.5) && (odr <= 25.0)) {
+		reg = BMM350_DATA_RATE_25HZ;
+	} else if ((odr > 25.0) && (odr <= 50.0)) {
+		reg = BMM350_DATA_RATE_50HZ;
+	} else if ((odr > 50.0) && (odr <= 100.0)) {
+		reg = BMM350_DATA_RATE_100HZ;
+	} else if ((odr > 100.0) && (odr <= 200.0)) {
+		reg = BMM350_DATA_RATE_200HZ;
+	} else if ((odr > 200.0) && (odr <= 400.0)) {
+		reg = BMM350_DATA_RATE_400HZ;
+	} else if (odr > 400.0) {
+		reg = BMM350_DATA_RATE_400HZ;
+	}
+	return reg;
+}
+/*!
+ * @brief This API sets the ODR and averaging factor.
+ */
+int8_t bmm350_set_odr_performance(enum bmm350_data_rates odr,
+				  enum bmm350_performance_parameters performance,
+				  const struct device *dev)
+{
+	/* Variable to store the function result */
+	int8_t rslt = 0x00;
+	/* Variable to get PMU command */
+	uint8_t reg_data = 0;
+	enum bmm350_performance_parameters performance_fix = performance;
+
+	if (rslt == BMM350_OK) {
+		/* Reduce the performance setting when too high for the chosen ODR */
+		if ((odr == BMM350_DATA_RATE_400HZ) && (performance >= BMM350_AVERAGING_2)) {
+			performance_fix = BMM350_NO_AVERAGING;
+		} else if ((odr == BMM350_DATA_RATE_200HZ) && (performance >= BMM350_AVERAGING_4)) {
+			performance_fix = BMM350_AVERAGING_2;
+		} else if ((odr == BMM350_DATA_RATE_100HZ) && (performance >= BMM350_AVERAGING_8)) {
+			performance_fix = BMM350_AVERAGING_4;
+		}
+
+		/* ODR is an enum taking the generated constants from the register map */
+		reg_data = ((uint8_t)odr & BMM350_ODR_MSK);
+		/* AVG / performance is an enum taking the generated constants from the register map
+		 */
+		reg_data = BMM350_SET_BITS(reg_data, BMM350_AVG, (uint8_t)performance_fix);
+		/* Set PMU command configurations for ODR and performance */
+		rslt = bmm350_reg_write(dev, BMM350_REG_PMU_CMD_AGGR_SET, reg_data);
+		LOG_DBG("odr index %d odr_reg_data 0x%x", odr, reg_data);
+
+		if (rslt == BMM350_OK) {
+			/* Set PMU command configurations to update odr and average */
+			reg_data = BMM350_PMU_CMD_UPD_OAE;
+			/* Set PMU command configuration */
+			rslt = bmm350_reg_write(dev, BMM350_REG_PMU_CMD, reg_data);
+			if (rslt == BMM350_OK) {
+				k_usleep(BMM350_UPD_OAE_DELAY);
+			}
+		}
+	}
+
+	return rslt;
+}
+static int set_mag_odr_osr(const struct device *dev, const struct sensor_value *odr,
+			   const struct sensor_value *osr)
+{
+	uint8_t odr_bits = 0x00;
+
+	if (odr) {
+		odr_bits = acc_odr_to_reg(odr);
+		if (bmm350_set_odr_performance((enum bmm350_data_rates)odr_bits, BMM350_AVERAGING_2,
+					       dev) < 0) {
+			LOG_ERR("bmm350_set_odr_performance failed");
+			return -EIO;
+		}
+	}
+	if (osr) {
+		if (osr->val1 == 0) {
+			if (bmm350_set_powermode(BMM350_SUSPEND_MODE, dev) < 0) {
+				LOG_ERR("bmm350_set_powermode suspend failed");
+				return -EIO;
+			}
+
+		} else {
+			if (bmm350_set_powermode(BMM350_NORMAL_MODE, dev) < 0) {
+				LOG_ERR("bmm350_set_powermode normal failed");
+				return -EIO;
+			}
+		}
+	}
```

Comment:
The oversampling rate is not truly being set here, it also may need to check that the oversampling rate multiplied by the output data rate does not exceed the 'max input data' rate of 400Hz so the application doesn't configure an invalid configuration

Reply:
now I understand, you want to configure the averaging mode here. I can add that.